### PR TITLE
[Core] Support timeout for gRPC methods.

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

Currently, every Ray. gRPC methods have infinite timeout. This is risky when we have high load, backpressured methods, or having a "blocking calls". 

Please refer to the issue for more details.

This PR refactors the code to support per-RPC-method timeout_ms.

This also converts gRPC timeout error code to `TimeOutError` Ray status code instead of `IOError`. This will help us having infinite retry when timeout happens from this code when timeout happens. cc @lixin-wei https://github.com/ray-project/ray/blob/722428a657158b771824de9eb63d468661925259/src/ray/rpc/gcs_server/gcs_rpc_client.h#L64

## Related issue number

https://github.com/ray-project/ray/issues/20733

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/rpc/grpc_client.h [hunk 1]
```diff
@@ -84,13 +85,16 @@ class GrpcClient {
...
+  /// \param[in] method_timeout_ms The timeout of the RPC method in ms.
+  /// -1 means it will use the default timeout configured for the handler.
   ///
   /// \return Status.
   template <class Request, class Reply>
   void CallMethod(
       const PrepareAsyncFunction<GrpcService, Request, Reply> prepare_async_function,
       const Request &request, const ClientCallback<Reply> &callback,
-      std::string call_name = "UNKNOWN_RPC") {
+      std::string call_name = "UNKNOWN_RPC", int64_t method_timeout_ms = -1) {
```
Comment (lixin-wei) [inline src/ray/rpc/grpc_client.h:97]: where is `method_timeout_ms` used in this method? I can't find any.

File: src/ray/rpc/gcs_server/gcs_rpc_client.h [hunk 2]
```diff
@@ -46,7 +46,8 @@ class Executor {
 };
 
 // Define a void GCS RPC client method.
-#define VOID_GCS_RPC_CLIENT_METHOD(SERVICE, METHOD, grpc_client, SPECS)                \
+#define VOID_GCS_RPC_CLIENT_METHOD(SERVICE, METHOD, grpc_client, method_timeout_ms,    \
```
Comment (lixin-wei) [inline src/ray/rpc/gcs_server/gcs_rpc_client.h:49]: where is `method_timeout_ms` used in this macro? Seems you hard-coded it as -1 below?

**Review by lixin-wei — CHANGES_REQUESTED** [review]

**Comment by lixin-wei** [comment]
BTW, what's the default timeout of gRPC itself? Do you know how to configure it?

File: src/ray/rpc/grpc_client.h (hunk 1 above)
Comment (rkooo567) [reply src/ray/rpc/grpc_client.h:97]: oh yeah good catch. Let me add them properly. Will add an unit test too

**Comment by rkooo567** [comment]
> BTW, what's the default timeout of gRPC itself? Do you know how to configure it? I think it'll be better if we can pass this timeout into gRPC.

We are migrating iteratively. By the fact that we have some long polling requests, I am pretty sure we don't have the timeout configured. You can set the global timeout, but it is too drastic change that's not safe.

The migration path is starting from (1) making sure blocking calls are correctly handled. (2) identify long polling and exclude them from timeout (3) and then migrate to general timeout. Not sure when we should do (3) (as mentioned this can be pretty drastic change since we don't have robust network failure logic)

**Comment by lixin-wei** [comment]
I think we do need to pass this timeout into gRPC. Or when GCS took a long time to handle a request, the request still won't be timed out.

I think we should add two params: `max_retry_times` and `timeout_ms`. `max_retry_times` controls how much time we'll retry after gRPC returns a non-OK status. `timeout_ms` controls the timeout of each gRPC request.

**Comment by rkooo567** [comment]
Hmm What you are saying is the impl detail of (3) right? We can discuss it when we work on that part. Timeout + retry only works when handlers are idempotent, otherwise, it can cause undefined behaviors. So we shouldn’t drastically migrate all at once.

**Comment by lixin-wei** [comment]
Nope I'm talking about 1). 
If you don't pass timeout into gRPC. How do you make RPC time out?

**Comment by rkooo567** [comment]
Oh not passing that is just a mistake. We already have timeout implemented, so we just need to pass the arg

**Comment by mwtian** [comment]
For the discussion above, IIIUC @lixin-wei is asking why we cannot pass per-call timeout / deadline, e.g. specifying a different timeout for each `PingTimeout` call. It might be useful to make it clear that this PR adds support for configuring per-handler timeout, in addition to the global client RPC timeout that was never feasible to be set. But it does not support per-call timeout.

I think this PR is a step in the right direction, if it can help avoiding the incident earlier. But the GRPC client wrapper should eventually allow specifying timeout or passing in `ClientContext` per call. Otherwise RPCs in Ray will not be able to support more complex use cases or user specified timeout well.

**Comment by rkooo567** [comment]
@mwtian I don't see the use case of per-call timeout right now, but I think adding support is trivial once we need it. We can consider that once the complex use case comes up!

**Review by lixin-wei — APPROVED** [review]
Looks good to me if CI passed.

**Comment by rkooo567** [comment]
Part 1 of https://github.com/ray-project/ray/issues/20899

**Pull request merged**